blog/src/a-base32-checksum.md

```diff
@@ -0,0 +1,1241 @@
+the primitive.
+
+- The **cardinal** is the number of elements in the finite field. It is
+  therefore also the number of possible checksums. For a single-character check
+  digit, it must be smaller than the size of your alphabet.
+- The **primitive** does not matter all that much. The most significant part
+  about it is that you need to compute one. Unfortunately, brute-force is usually
+  the simplest way to find some.
+
+There are two possible cardinals: *prime numbers* and *prime powers*. Prime
+cardinals are neat, because operations are scalar. On the other hand, additions
+and multiplications on prime power cardinal fields are matricial.
+
+I first wanted a single-character checksum. In order to use all the bits that
+this checksum had at its disposal, I needed it to support the full base32
+alphabet. That meant the cardinal needed to be 32. Sadly, 32 is not a prime; so
+it needed to be a prime power. If 2 is the prime, 2⁵ is the cardinal.
+
+Since the cardinal is not a prime, P is a matrix. Here, its size is 5×5.
+To build it, we can pick any order-5 [primitive polynomial][]; I chose 1+x²+x⁵.
+To get the companion matrix, set a diagonal of 1s, and set the rightmost column
+with the polynomial coefficients, starting with x⁰:
+
+    ⎛ 0 0 0 0 1 ⎞
+    ⎜ 1 0 0 0 0 ⎟
+    ⎜ 0 1 0 0 1 ⎟
+    ⎜ 0 0 1 0 0 ⎟
+    ⎝ 0 0 0 1 0 ⎠
+
+Each aᵢ term is a representation of a base32 character as a GF(2⁵) (finite field
+of order 2⁵) polynome: a is 0, which we represent as (0 0 0 0 0); b is 1 or (0 0
+0 0 1); f is 1+x² or (0 0 1 0 1).
+
+Validating a check character is straightforward: compute Σ aᵢ·Pⁱ; verify that
+the result is a zero vector. It does involve writing custom matrix
+multiplication and addition primitives to ensure that they are all computed
+modulo the cardinal.
+
+Unsurprisingly, the performance is not all that great:
+
+    $ node test/perf
+    10000 runs.
+    base32check1: 243.500ms
+
+There is a [neat trick][ffcomp] to speed up computation on a finite field where
+the cardinal is a power of 2: instead of implementing vectors as arrays, use
+integers whose binary sequence is the list of polynome coefficients. For
+instance, 1+x² (which represents f) is `0b00101`, or 5.
+
+Then **addition is XOR**, and matrix multiplication is a conditional XOR on the
+successive bits of the current row. With aᵢ and bᵢ vectors of size 5:
+
+    ⎛ a₀ ⎞   ⎛ b₀ ⎞   ⎛ Σ a₀[i]×bᵢ ⎞
+    ⎜ a₁ ⎟   ⎜ b₁ ⎟   ⎜ Σ a₁[i]×bᵢ ⎟
+    ⎜ a₂ ⎟ × ⎜ b₂ ⎟ = ⎜ Σ a₂[i]×bᵢ ⎟
+    ⎜ a₃ ⎟   ⎜ b₃ ⎟   ⎜ Σ a₃[i]×bᵢ ⎟
+    ⎝ a₄ ⎠   ⎝ b₄ ⎠   ⎝ Σ a₄[i]×bᵢ ⎠
+
+It is 7× faster.
+
+    $ node test/perf
+    10000 runs.
+    base32check1: 34.120ms
+
+Now, how do you compute the check character?
+
+First, compute S = Σ aᵢ·Pⁱ for the data you have.
+Then, we need to solve S + c·P<sup>n+1</sup> = 0 for c.
+
+Let's compute c·P<sup>n+1</sup> = -S.
+Opposites are equal in GF(2ᵏ), so -S = S.
+
+The second insight is that, since a primitive element is a generator of
```

Comment:
Are you talking about the second insight of the 2016 paper?

Reply:
Ah, no. You need two insights to solve the equation. The first trick is that computing opposites is trivial because it is an identity operation; the second is this one.

I will switch to this formulation:

```markdown
Now, we need to inverse P<sup>n+1</sup>.
To do that efficiently, we need the following insight: since a primitive element […]
```